cmds/core/free/free.go

```diff
@@ -140,109 +133,138 @@ func humanReadableValue(value uint64) string {
 // depending on whether FreeConfig specifies a human-readable format or a
 // specific unit
-func formatValueByConfig(value uint64, config *FreeConfig) string {
-	if config.HumanOutput {
+func (c *cmd) formatValueByConfig(value uint64) string {
+	if c.human {
 		return humanReadableValue(value)
 	}
 	// units and decimal part are not printed when a unit is explicitly specified
-	return fmt.Sprintf("%v", value>>config.Unit)
+	return fmt.Sprintf("%v", value>>c.unit)
 }
 
-// Free prints physical memory and swap space information. The fields will be
-// expressed with the specified unit (e.g. KB, MB)
-func Free(config *FreeConfig) error {
-	m, err := meminfo()
+func main() {
+	flag.Parse()
+	o := options{human: *humanOutput, bytes: *inBytes, kbytes: *inKB, mbytes: *inMB, gbytes: *inGB, tbytes: *inTB, json: *toJSON}
+	cmd, err := command(os.Stdout, o)
 	if err != nil {
 		log.Fatal(err)
 	}
-	mmi, err := getMainMemInfo(m, config)
-	if err != nil {
-		return err
-	}
-	si, err := getSwapInfo(m, config)
-	if err != nil {
-		return err
-	}
-	mi := MemInfo{Mem: *mmi, Swap: *si}
-	if config.ToJSON {
-		jsonData, err := json.Marshal(mi)
-		if err != nil {
-			return err
-		}
-		fmt.Println(string(jsonData))
-	} else {
-		fmt.Printf("              total        used        free      shared  buff/cache   available\n")
-		fmt.Printf("%-7s %11v %11v %11v %11v %11v %11v\n",
-			"Mem:",
-			formatValueByConfig(mmi.Total, config),
-			formatValueByConfig(mmi.Used, config),
-			formatValueByConfig(mmi.Free, config),
-			formatValueByConfig(mmi.Shared, config),
-			formatValueByConfig(mmi.Buffers+mmi.Cached, config),
-			formatValueByConfig(mmi.Available, config),
-		)
-		fmt.Printf("%-7s %11v %11v %11v\n",
-			"Swap:",
-			formatValueByConfig(si.Total, config),
-			formatValueByConfig(si.Used, config),
-			formatValueByConfig(si.Free, config),
-		)
+	if err = cmd.run(); err != nil {
+		log.Fatal(err)
 	}
-	return nil
 }
 
-// validateUnits checks that only one option of -b, -k, -m, -g, -t or -h has been
-// specified on the command line
-func validateUnits() bool {
+type cmd struct {
+	stdout io.Writer
+	unit   unit
+	human  bool
+	toJSON bool
+}
+
+type options struct {
+	human  bool
+	bytes  bool
+	kbytes bool
+	mbytes bool
+	gbytes bool
+	tbytes bool
+	json   bool
+}
+
+func command(stdout io.Writer, o options) (*cmd, error) {
+	// validateUnits checks that only one option of -b, -k, -m, -g, -t or -h has been
+	// specified on the command line
 	count := 0
-	if *inBytes {
+	if o.bytes {
 		count++
 	}
-	if *inKB {
+	if o.kbytes {
 		count++
 	}
-	if *inMB {
+	if o.mbytes {
 		count++
 	}
-	if *inGB {
+	if o.gbytes {
 		count++
 	}
-	if *inTB {
+	if o.tbytes {
 		count++
 	}
-	if *humanOutput {
+	if o.human {
 		count++
 	}
 	if count > 1 {
-		return false
+		return nil, errMultipleUnits
 	}
-	return true
-}
 
-func main() {
-	flag.Parse()
-	if !validateUnits() {
-		log.Fatal("Options -k, -m, -g, -t and -h are mutually exclusive")
+	c := &cmd{
```

Comment:
The problem is that k, m, h, t, and h are radiobuttons, and the golang flags package has no good way to do that. I've tried and the code is just unbelievably awkward. The messy code in this is from the original, but at some point, it would be nice to do better. 

Just to count  the number of options set, I think we could least do this:
func countTrue(b ... boolean) int {
var count int
for _, v := range b {
count++
}
return count
}

and then in that awkard bunch of ifs:
if countTrue(o.bytes, o.mbytes, o.gbytes, o.tbytes, o.human) > 1 {
return errMultipleUnits
}

Reply:
thx, updated